### PR TITLE
ENH:stats: Add multivariate Wasserstein distance as a separate function

### PR DESCRIPTION
**Pull request by yagizolmez**
#17473 extended `wasserstein_distance` to 2D, which was set to be released with the version `1.12.0`. Later discussions showed that it would be better to keep `wasserstein_distance` as 1D. Therefore #19881 reverted #17473. This PR builds on #17473 and creates another function, `wasserstein_distance_nd`, to handle the 2D case, as stipulated by #19881.

@com3dian, @mdhaber I would be happy if you could take a look.

### CODE REVIEW

File: scipy/stats/_stats_py.py
Comment (VascoSch92) [inline scipy/stats/_stats_py.py:43]: can you rewrite as

```python
from scipy._lib._util import (
    check_random_state, 
    MapWrapper, 
    _get_nan,
    rng_integers,
    _rename_parameter,
    _contains_nan,
    AxisError,
)
```

sorry it has nothing to do with your PR but it is really difficult to read :-)

Comment (lucascolley) [reply scipy/stats/_stats_py.py:43]: Hi @VascoSch92, we generally avoid touching lines which are unrelated to the PR. For style-only changes like this, we prefer to deal with them only when the lines are already being touched, or in wholesale changes. This makes reviewing individual PRs easier and keeps the git blame history more clean. See https://scipy.github.io/devdocs/dev/contributor/pep8.html#pep8-and-scipy.

Comment (VascoSch92) [reply scipy/stats/_stats_py.py:43]: ah ok ;-) 

sorry i was just reading the PR and i saw that.

**Comment by mdhaber** [comment]
Thanks @yagizolmez. All new functions need approval from the SciPy [mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/). Would you send a message to the list proposing to add this function asking for feedback on this PR?

**Comment by yagizolmez** [comment]
@mdhaber Done!

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10403]:
```suggestion
    Compute the Wasserstein-1 distance between two 1D discrete distributions.
```
@yagizolmez Could you also add the relevant docstring changes from gh-17473 here? Please read through all the docstring changes (including examples) there and comment about which changes are still relevant to `wasserstein_distance` and which only apply to `wassertein_distance_nd`.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10480]: Please also consider whether the new input validation there would still be appropriate. (Perhaps it is performed in _cdf_distance, but if not, we should perform the checks and add tests of the input validation.)

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10264]: PEP8 recommends two blank lines between function definitions, so please undo this change.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7441]: Please create a new class `TestWassersteinDistanceND` and add the new tests there instead of interspersing the two.

**Review by mdhaber — COMMENTED** [review]
Some preliminary comments to get started.

Also, I am going to review this mostly by comparing what you are adding here vs what was added in gh-17473 to see that they are identical. If you know you made changes in some places (other than, say, whitespace or adding comments), please comment about those changes here in the PR (not the code).

File: scipy/stats/_stats_py.py
Comment (yagizolmez) [reply scipy/stats/_stats_py.py:10480]: I think it would be appropriate. I have not changed the code in `_stats_py.py`. I have only modified the docs and tests.

File: scipy/stats/_stats_py.py
Comment (yagizolmez) [reply scipy/stats/_stats_py.py]: Done!

File: scipy/stats/_stats_py.py
Comment (yagizolmez) [reply scipy/stats/_stats_py.py]: Done!

File: scipy/stats/tests/test_stats.py
Comment (yagizolmez) [reply scipy/stats/tests/test_stats.py]: Done!

File: scipy/stats/_stats_py.py
Comment (yagizolmez) [reply scipy/stats/_stats_py.py]: @mdhaber I think there has been a problem with CI in the past few days. I don't undestand why it failed some checks after commiting this suggestion. I am having the same issue in #20058.

Comment (mdhaber) [reply scipy/stats/_stats_py.py]: If you commit again while CI is running, the check is cancelled and shows up as a failure.
As a new contributor, most jobs don't run without approval. I approved after your last commit, so please don't commit again until that's all done.

Comment (yagizolmez) [reply scipy/stats/_stats_py.py]: @mdhaber Thank you so much! Can you do the same for #20058?

**Comment by yagizolmez** [comment]
@mdhaber Some tests are failing, but I didn't see any relation to this PR in the logs. Is this normal?

**Comment by mdhaber** [comment]
Unfortunately, yes, it is normal!

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10415]: We did some work on the parameter documentation in gh-17473 that should be preserved. Please commit this and remove the existing Parameters section.

```suggestion

    Parameters
    ----------
    u_values : 1d array_like
        A sample from a probability distribution or the support (set of all
        possible values) of a probability distribution. Each element along
        axis 0 is an observation or possible value.

    v_values : 1d array_like
        A sample from or the support of a second distribution.

    u_weights, v_weights : 1d array_like, optional
        Weights or counts corresponding with the sample or probability masses
        corresponding with the support values. Sum of elements must be positive
        and finite. If unspecified, each value is assigned the same weight.
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10194]:
```suggestion
    Compute the Wasserstein-1 distance between two N-D discrete distributions.
```
?

The _arrays_ are 2D; that second dimension of the array is used to represent an arbitrary number of dimensions of the distribution. That's why it's called `wasserstein_distance_nd`, after all.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10239]:
```suggestion
    :math:`\mathbb{R}^n \times \mathbb{R}^n` whose marginals are :math:`u` and
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10474]:
```suggestion
    wasserstein_distance_nd: Compute the Wasserstein-1 distance between two N-D
```

**Review by mdhaber — CHANGES_REQUESTED** [review]
@yagizolmez this looks pretty good! I checked that everything was copied over and that there are no other unnecessary changes. There were just a few things I noticed; specific suggestions inline. Please look for other cases in which 2-D should be N-D.

Besides that, it would be great if @com3dian could take a quick look - mostly at whether the references in the two functions are still correct, and also if the the section "Given two probability mass function..." has been converted to the N-D case appropriately. 

(Otherwise, I compared the diffs of the new `wasserstein_distance` and `wasserstein_distance_nd` against `wasserstein_distance` from gh-17473, and I compared the new `TestWassersteinDistanceND` and `TestWassersteinDistance` against `TestWassersteinDistance` from gh-17473. Everything was essentially line-for-line identical with only the necessary changes, like all 1-D parts removed from the new N-D sections, and all N-D parts removed from the new 1-D sections.)

File: scipy/stats/_stats_py.py
Comment (yagizolmez) [reply scipy/stats/_stats_py.py:10415]: @mdhaber In the description of `u_values`, I think we could change the sentence:

`Each element along axis 0 is an observation or possible value.`

`u_values` must have only one axis, so specifying `axis 0` is confusing. We could do:

`Each element is an observation or possible value.`

Please let me know what you think. If you agree, I will commit your suggestion and change this sentence in a subsequent commit.

Comment (mdhaber) [reply scipy/stats/_stats_py.py:10415]: Ah yes, of course. Thanks, please change that.

**Comment by com3dian** [comment]
@mdhaber @yagizolmez , I just checked the doc string and it looks okay. My only suggestions are  1. consider remove reference [3] in `wasserstein_distance_nd` function as it is about equivalence between CDF distance and and  2. reverse the order of reference 4 and 5 in `wasserstein_distance_nd` function.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10435]:
```suggestion
    Given two 1D probability mass functions, :math:`u` and :math:`v`, the first
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10468]: Link is not rendering correctly, so I suppose it needs to be all on one line. Feel free to make it `"""  # noqa: e501` at the end of the docstring if the line is too long.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:10465]: Would be good to refer to this in the text if it's being cited. It could go after the first sentence at the top.

**Review by mdhaber — APPROVED** [review]
LGTM after these little tweaks are made. The reference suggestions apply to the `_nd` function, too. I'll plan waiting until Wednesday to merge, since that will be a week after your email to the mailing list went out.

File: scipy/stats/_stats_py.py
Comment (yagizolmez) [reply scipy/stats/_stats_py.py]: Linter didn't give an error for me locally. Let's see how it goes here.

File: scipy/stats/tests/test_stats.py
Comment (j-bowhay) [inline scipy/stats/tests/test_stats.py:7569]: In the new tests you have added could we use [assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) instead of `assert_almost_equal`

Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:7569]: There weren't any new tests here, but it looks like those made it through from the original PR. Yes, please change all the Wasserstein distance tests to `assert_allclose`.

Comment (yagizolmez) [reply scipy/stats/tests/test_stats.py:7569]: > I don't think there were any new tests here (right?), but it looks like those made it through from the original PR.

No new tests were added to `TestWassersteinDistance` covering the 1-D case in the original PR, but some of the inline comments were modified- those made through.

>  Yes, please change all the Wasserstein distance tests to `assert_allclose`.

I will.

**Comment by yagizolmez** [comment]
I apologize for the unclear commit message. I was trying to write:

> Changed instances of `assert_almost_equal` to `assert_allclose`

**Comment by yagizolmez** [comment]
@mdhaber Should we merge, so that we can proceed with #19876 and #19879?

**Comment by mdhaber** [comment]
Yup, well, nobody has dissented, but I was hoping we could get a few +1s. 
@j-bowhay @rlucas7 I know you didn't review in detail, but does it look good to you to have this function?

**Comment by mdhaber** [comment]
@yagizolmez I'm going to rewrite the history here, but I'll save a backup branch for you in case you want it. Then you'll be able to rebase those other PRs on this branch.

**Comment by yagizolmez** [comment]
@mdhaber Thank you. Do I need to do anything?

**Comment by lucascolley** [comment]
> @j-bowhay @rlucas7 I know you didn't review in detail, but does it look good to you to have this function?

I suppose this tag was intended to be me, since I commented above. Sounds good to me :)

**Comment by mdhaber** [comment]
You too, @lucascolley, but Jake commented, too. Sorry I left you off.

**Comment by mdhaber** [comment]
@yagizolmez Nope, but now that I've rebased, you can pretend this branch is `main` and rebase those other PRs on this. I'll merge when CI finishes.

**Comment by mdhaber** [comment]
Thanks @yagizolmez et al!

**Pull request merged**
